src/Agent.Plugins/TestFilePublisher/TestFilePublisher.cs

```diff
@@ -0,0 +1,111 @@
+            _pipelineConfig = pipelineConfig;
+            _logger = logger;
+            _telemetry = telemetry;
+        }
+
+        public TestFilePublisher(VssConnection vssConnection, PipelineConfig pipelineConfig, TraceListener traceListener,
+            ITraceLogger logger, ITelemetryDataCollector telemetry, ITestFileFinder testFileFinder,
+            ITestResultParser testResultParser, ITestRunPublisher testRunPublisher)
+        : this(vssConnection, pipelineConfig, traceListener, logger, telemetry)
+        {
+            _testFileFinder = testFileFinder;
+            _testResultParser = testResultParser;
+            _testRunPublisher = testRunPublisher;
+        }
+
+        public async Task InitializeAsync()
+        {
+            await Task.Run(() => Initialize());
+        }
+
+        public async Task PublishAsync()
+        {
+            var testResultFiles = new List<string>();
+            IList<TestRunData> testData;
+
+            var testRunContext = new TestRunContextBuilder("Auto Published Test Run")
+                .WithBuildId(_pipelineConfig.BuildId)
+                .WithBuildUri(_pipelineConfig.BuildUri)
+                .Build();
+
+            using (new SimpleTimer(TelemetryConstants.FindTestFilesAsync, _logger, TimeSpan.FromSeconds(60),
+                new TelemetryDataWrapper(_telemetry, TelemetryConstants.FindTestFilesAsync)))
+            {
+                testResultFiles.AddRange(await FindTestFilesAsync());
+                _logger.Info($"Number of files found with matching pattern {testResultFiles.Count}");
+                _telemetry.AddAndAggregate("NumberOfTestFilesFound", testResultFiles.Count);
+            }
+
+            if (!testResultFiles.Any())
+            {
+                _logger.Info("No test files are found");
+                return;
+            }
+
+            using (new SimpleTimer(TelemetryConstants.ParseTestResultFiles, _logger, TimeSpan.FromSeconds(60),
+                new TelemetryDataWrapper(_telemetry, TelemetryConstants.ParseTestResultFiles)))
+            {
+                testData = _testResultParser.ParseTestResultFiles(testRunContext, testResultFiles).GetTestRunData();
+                _logger.Info($"Successfully parsed {testData.Count} files");
+                _telemetry.AddAndAggregate("NumberOfTestFilesRead", testData.Count);
+            }
+
+            using (new SimpleTimer(TelemetryConstants.PublishTestRunDataAsync, _logger, TimeSpan.FromSeconds(60),
+                new TelemetryDataWrapper(_telemetry, TelemetryConstants.PublishTestRunDataAsync)))
+            {
+                var publishedRuns = await _testRunPublisher.PublishTestRunDataAsync(testRunContext, _pipelineConfig.ProjectName, testData, new PublishOptions(),
+                    new CancellationToken());
+                _logger.Info($"Successfully published {publishedRuns.Count} runs");
+                _telemetry.AddAndAggregate("NumberOfTestRunsPublished", publishedRuns.Count);
```

Comment:
please log test run ids of each test run published and also send them to telemetry